### PR TITLE
Add lint for unused locals

### PR DESCRIPTION
**Pull request by jryans**
TypeScript supports checking for "unused locals", which actually includes
locals, private fields, private methods, etc. If a private thing is actually
needed as part of the API surface, you can change it to public to resolve the
error.

This seems like a good check to have, as it highlights unused bits of code that
perhaps got lost while refactoring. This changes our config to reject them and
fixes up a few existing cases.

See also https://github.com/matrix-org/matrix-js-sdk/pull/2223

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

Preview: https://pr8007--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

File: src/components/structures/RoomView.tsx [hunk 1]
```diff
@@ -1681,7 +1681,7 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
      *
      * We pass it down to the scroll panel.
      */
-    private handleScrollKey = ev => {
+    public handleScrollKey = ev => {
```
Comment (t3chguy) [inline src/components/structures/RoomView.tsx:1684]: Any chance you could fix the LIV side of this? This should've been caught without unused locals lint.

```diff
Index: src/components/structures/LoggedInView.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/components/structures/LoggedInView.tsx b/src/components/structures/LoggedInView.tsx
--- a/src/components/structures/LoggedInView.tsx	(revision 782ce016d1ad137163e2b47f221e05b71eafd7d8)
+++ b/src/components/structures/LoggedInView.tsx	(date 1646825378108)
@@ -57,7 +57,7 @@
 import AudioFeedArrayForCall from '../views/voip/AudioFeedArrayForCall';
 import { OwnProfileStore } from '../../stores/OwnProfileStore';
 import { UPDATE_EVENT } from "../../stores/AsyncStore";
-import RoomView from './RoomView';
+import RoomView, { RoomView as RoomViewComponent } from './RoomView';
 import ToastContainer from './ToastContainer';
 import MyGroups from "./MyGroups";
 import UserView from "./UserView";
@@ -139,7 +139,7 @@
     static displayName = 'LoggedInView';
 
     protected readonly _matrixClient: MatrixClient;
-    protected readonly _roomView: React.RefObject<any>;
+    protected readonly _roomView: React.RefObject<RoomViewComponent>;
     protected readonly _resizeContainer: React.RefObject<HTMLDivElement>;
     protected readonly resizeHandler: React.RefObject<HTMLDivElement>;
     protected layoutWatcherRef: string;

**Review by t3chguy — APPROVED** [review]
Looks awesome otherwise

:shipit:

File: src/components/structures/RoomView.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/structures/RoomView.tsx:1684]: Or actually, better yet, ditch the withMatrixClientHOC in RoomView.tsx as RoomView uses React Context and never uses `props.mxClient`

Comment (t3chguy) [reply src/components/structures/RoomView.tsx:1684]: Something akin to

```diff
Index: src/components/structures/RoomView.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/components/structures/RoomView.tsx b/src/components/structures/RoomView.tsx
--- a/src/components/structures/RoomView.tsx	(revision 782ce016d1ad137163e2b47f221e05b71eafd7d8)
+++ b/src/components/structures/RoomView.tsx	(date 1646826128031)
@@ -57,7 +57,7 @@
 import RightPanelStore from "../../stores/right-panel/RightPanelStore";
 import { haveTileForEvent } from "../views/rooms/EventTile";
 import RoomContext, { TimelineRenderingType } from "../../contexts/RoomContext";
-import MatrixClientContext, { MatrixClientProps, withMatrixClientHOC } from "../../contexts/MatrixClientContext";
+import MatrixClientContext from "../../contexts/MatrixClientContext";
 import { E2EStatus, shieldStatusForRoom } from '../../utils/ShieldUtils';
 import { Action } from "../../dispatcher/actions";
 import { IMatrixClientCreds } from "../../MatrixClientPeg";
@@ -119,7 +119,7 @@
     debuglog = logger.log.bind(console);
 }
 
-interface IRoomProps extends MatrixClientProps {
+interface IRoomProps {
     threepidInvite: IThreepidInvite;
     oobData?: IOOBData;
 
@@ -219,7 +219,7 @@
 }
 
 @replaceableComponent("structures.RoomView")
-export class RoomView extends React.Component<IRoomProps, IRoomState> {
+export default class RoomView extends React.Component<IRoomProps, IRoomState> {
     private readonly dispatcherRef: string;
     private readonly roomStoreToken: EventSubscription;
     private settingWatchers: string[];
@@ -1681,7 +1681,7 @@
      *
      * We pass it down to the scroll panel.
      */
-    private handleScrollKey = ev => {
+    public handleScrollKey = ev => {
         let panel;
         if (this.searchResultsPanel.current) {
             panel = this.searchResultsPanel.current;
@@ -2184,6 +2184,3 @@
         );
     }
 }
-
-const RoomViewWithMatrixClient = withMatrixClientHOC(RoomView);
-export default RoomViewWithMatrixClient;
Index: src/components/structures/LoggedInView.tsx
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/components/structures/LoggedInView.tsx b/src/components/structures/LoggedInView.tsx
--- a/src/components/structures/LoggedInView.tsx	(revision 782ce016d1ad137163e2b47f221e05b71eafd7d8)
+++ b/src/components/structures/LoggedInView.tsx	(date 1646826128036)
@@ -139,7 +139,7 @@
     static displayName = 'LoggedInView';
 
     protected readonly _matrixClient: MatrixClient;
-    protected readonly _roomView: React.RefObject<any>;
+    protected readonly _roomView: React.RefObject<RoomView>;
     protected readonly _resizeContainer: React.RefObject<HTMLDivElement>;
     protected readonly resizeHandler: React.RefObject<HTMLDivElement>;
     protected layoutWatcherRef: string;

Comment (jryans) [reply src/components/structures/RoomView.tsx:1684]: Ah sure, sounds like a good tweak! 😄

**Pull request merged**